ml/pipelines/binary.py

```diff
@@ -149,48 +152,54 @@ def _run_all_models(self, refactoring, refactoring_name, dataset, features, scal
         for model in self._models_to_run:
             model_name = model.name()
             if TEST:
-                model_name += " test"
+                model_name += " val"
             try:
                 log("\nBuilding Model {}".format(model.name()))
                 self._start_time()
-                test_scores, test_results, model_to_save = self._run_single_model(model, x, y, x_train, x_tests, y_train, y_tests, db_ids)
+                features, val_scores, val_results, model_to_save = self._run_single_model(model, x, y, x_train, x_val_list, y_train, y_val_list, db_ids)
 
-                # log test scores
-                formatted_results = format_results_single_run(dataset, refactoring_name, test_names, model_name, test_scores["precision"],
-                                            test_scores["recall"], test_scores['accuracy'], test_scores['tn'],
-                                            test_scores['fp'], test_scores['fn'], test_scores['tp'],
-                                              model_to_save, features)
+                # log val scores
+                formatted_results = format_results_single_run(dataset, refactoring_name, val_names, model_name, val_scores["precision"],
+                                                              val_scores["recall"], val_scores['accuracy'], val_scores['tn'],
+                                                              val_scores['fp'], val_scores['fn'], val_scores['tp'], val_scores["permutation_importance"],
+                                                              model_to_save, features)
                 log(formatted_results)
 
                 # we save the best estimator we had during the search
                 # Also, store the predictions with labels and db_ids, to analyze them later
                 model.persist(dataset, refactoring_name, features, model_to_save, scaler,
-                              test_results, test_names, formatted_results)
+                              val_results, val_names, formatted_results)
                 self._finish_time(dataset, model, refactoring)
             except Exception as e:
                 log("An error occurred while working on refactoring " + refactoring_name + " model " + model.name()
-                       + " with datasets: " + str(test_names))
+                    + " with datasets: " + str(val_names))
                 log(str(e))
                 log(str(traceback.format_exc()))
 
-    def _run_single_model(self, model_def, x, y, x_train, x_tests, y_train, y_tests, db_ids):
+    def _run_single_model(self, model_def, X, y, x_train, x_val_list, y_train, y_val_list, db_ids):
         model = model_def.model()
 
         # perform the search for the best hyper parameters
         param_dist = model_def.params_to_tune()
         search = None
 
+        features, x_train = perform_feature_reduction(model, x_train, y_train)
```

Comment:
hmm, this is interesting. We now perform feature reduction inside the main training pipeline. Can we really use the same model here to do the feature reduction? Is that the right thing to do?

I feel we should use a simplified model to perform the feature reduction and not the specific model. And then give the data to the model we want to evaluate.

Otherwise, behind the scenes, we will be training the same model (maybe with a simplified non-tuned version of the model) that we test.

What do you think?

Reply:
Usually, we use [SelectFromModel](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectFromModel.html#sklearn.feature_selection.SelectFromModel) to select the features and this would often use the same classifier, but not always. Unfortunately, the architecture of our ml pipeline, makes this quite difficult, thus I had to move it where it is now.

Using a simplified estimator for training is quite redundant for us because we are already using very simple classifiers. (LR, RF, etc.)

Reply:
The example in there indeed shows them using the same model for feature selection and then training. 

@dvanderleij, @Dahny do you agree? (Let me know if you need more context to understand)

Reply:
I like this position for the feature reduction as it's just using the training set now, which is better practice in general.

I think it's fine to use the same model for the feature reduction as it's more relevant for the to be trained model. I am not sure how biased this will become, this will most likely also depend on the hyperparameter optimization.  But intuitively I would say that the potential bias here is most likely negligible.